### PR TITLE
When changing IV length invalidate previously set IV [3.1]

### PR DESCRIPTION
**Pull request by t8m**
Same as #22590 for 3.1 and 3.0 branches.

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

File: test/evp_extra_test.c
Comment (slontis) [inline test/evp_extra_test.c:5349]: Is .key != NULL clearer?

File: providers/implementations/ciphers/ciphercommon_ccm.c
Comment (slontis) [inline providers/implementations/ciphers/ciphercommon_ccm.c:114]: Is this  tested-  what happens with this line.... ```` if (!ctx->len_set && !ccm_set_iv(...)````)

File: providers/implementations/ciphers/cipher_aes_ocb.c
Comment (slontis) [inline providers/implementations/ciphers/cipher_aes_ocb.c:391]: This would only work if it was called before aes_ocb_init()? (Since update_iv() only works with IV_STATE_BUFFERED)

File: providers/implementations/ciphers/ciphercommon_ccm.c
Comment (slontis) [reply providers/implementations/ciphers/ciphercommon_ccm.c:114]: Same comment appllies to gcm

Comment (slontis) [reply providers/implementations/ciphers/ciphercommon_ccm.c:114]: i.e. Is the test falling over for the reason you think it is.

File: providers/implementations/ciphers/ciphercommon_ccm.c
Comment (DIEGOHJJACOBO) [inline providers/implementations/ciphers/ciphercommon_ccm.c:112]: providers/implementations/ciphers/cipher_aes_ocb.c

File: providers/implementations/ciphers/ciphercommon_gcm.c
Comment (DIEGOHJJACOBO) [inline providers/implementations/ciphers/ciphercommon_gcm.c:265]: providers/implementations/ciphers/cipher_aes_ocb.c

Comment (DIEGOHJJACOBO) [inline providers/implementations/ciphers/ciphercommon_gcm.c:265]: https://github.dev/DIEGOHJJACOBO

**Review by DIEGOHJJACOBO — APPROVED** [review]
DIEGOHJJACOBO/providers/implementations/ciphers/ciphercommon.c

File: test/evp_extra_test.c
Comment (t8m) [reply test/evp_extra_test.c:5349]: IMO no.

File: providers/implementations/ciphers/cipher_aes_ocb.c
Comment (t8m) [reply providers/implementations/ciphers/cipher_aes_ocb.c:391]: Yes, setting **a different** IV length must work only if you call aes_ocb_init() setting a new IV. Other cases must fail because the IV is effectively unset after you change the IV length.

File: providers/implementations/ciphers/ciphercommon_ccm.c
Comment (t8m) [reply providers/implementations/ciphers/ciphercommon_ccm.c:114]: I am sorry, it is totally unclear to me what do you mean. The test fails without the change.

**Comment by t8m** [comment]
Ping for second review

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
pauli's review does not count anymore because he's on leave of absence. Please @openssl/committers review.

**Comment by t8m** [comment]
Merged to the 3.1 and 3.0 branches. Thank you for the reviews.

**Pull request closed without merging**
